api/v1/_email/PKPEmailHandler.inc.php

```diff
@@ -0,0 +1,210 @@
+			'POST' => [
+				[
+					'pattern' => $this->getEndpointPattern(),
+					'handler' => [$this, 'create'],
+					'roles' => [ROLE_ID_SITE_ADMIN, ROLE_ID_MANAGER],
+				],
+			],
+			'PUT' => [
+				[
+					'pattern' => $this->getEndpointPattern() . '/{queueId}',
+					'handler' => [$this, 'process'],
+					'roles' => [ROLE_ID_SITE_ADMIN, ROLE_ID_MANAGER],
+				],
+			],
+		];
+		parent::__construct();
+	}
+
+	/**
+	 * @copydoc PKPHandler::authorize
+	 */
+	public function authorize($request, &$args, $roleAssignments) {
+		import('lib.pkp.classes.security.authorization.PolicySet');
+		$rolePolicy = new PolicySet(COMBINING_PERMIT_OVERRIDES);
+
+		import('lib.pkp.classes.security.authorization.RoleBasedHandlerOperationPolicy');
+		foreach ($roleAssignments as $role => $operations) {
+			$rolePolicy->addPolicy(new RoleBasedHandlerOperationPolicy($request, $role, $operations));
+		}
+		$this->addPolicy($rolePolicy);
+
+		return parent::authorize($request, $args, $roleAssignments);
+	}
+
+	/**
+	 * Create a jobs queue to send a bulk email to users in one or
+	 * more user groups
+	 *
+	 * @param ServerRequestInterface $slimRequest
+	 * @param APIResponse $response
+	 * @param array $args arguments
+	 * @return APIResponse
+	 */
+	public function create(ServerRequestInterface $slimRequest, APIResponse $response, array $args) {
+		$context = $this->getRequest()->getContext();
+		$contextId = $context->getId();
+
+		$requestParams = $slimRequest->getParsedBody();
+
+		$params = [];
+		foreach ($requestParams as $param => $val) {
+			switch ($param) {
+				case 'userGroupIds':
+					if (!is_array($val)) {
+						$val = strlen(trim($val))
+							? explode(',', $val)
+							: [];
+					}
+					$params[$param] = array_map('intval', $val);
+					break;
+				case 'body':
+				case 'subject':
+					$params[$param] = $val;
+					break;
+			}
+		}
+
+		$errors = [];
+		if (empty($params['body'])) {
+			$errors['body'] = [__('api.emails.400.missingBody')];
+		}
+
+		if (empty($params['subject'])) {
+			$errors['subject'] = [__('api.emails.400.missingSubject')];
+		}
+
+		if (empty($params['userGroupIds'])) {
+			$errors['userGroupIds'] = [__('api.emails.400.missingUserGroups')];
+		}
+
+		if ($errors) {
+			return $response->withJson($errors, 400);
+		}
+
+		$userGroupDao = DAORegistry::getDAO('UserGroupDAO');
+		foreach ($params['userGroupIds'] as $userGroupId) {
+			if (!$userGroupDao->contextHasGroup($contextId, $userGroupId)) {
+				return $response->withJson([
+						'userGroupIds' => [__('api.emails.403.notAllowedUserGroup')],
+					], 403);
+			}
+		}
+
+		// Only permit emails to be sent to active users in this context
+		$params['status'] = 'active';
+		$params['contextId'] = $contextId;
+
+		$userIds = Services::get('user')->getIds($params);
+		$subject = $params['subject'];
+		$body = $params['body'];
+		$fromEmail = $context->getData('contactName');
+		$fromName = $context->getData('contactEmail');
+		$queueId = 'email_' . uniqid();
```

Comment:
It's too bad queues can't have both names and IDs. I'm halfway tempted to propose that we just call this queue `email_notify` or something and have all batch email sending use the single queue. This would facilitate the fat-finger check I was suggesting earlier:
> Before we kick off a queueing operation (#1), we can check that the queue is currently empty (#3). If it is, we can go ahead.
> If it's not empty, we can ask the user if they want to flush the queue first (#4).

However, that would be a single site-wide queue, which is obviously bad -- so maybe this is the least-worst solution for now?

Reply:
Here the 
```
$fromEmail = $context->getData('contactName');
$fromName = $context->getData('contactEmail');
```
should probably the other way around

Reply:
Good catch @defstat 

> However, that would be a single site-wide queue, which is obviously bad -- so maybe this is the least-worst solution for now?

Yeah, this is a works-for-one-or-two-cases kind of approach. I think eventually we will need an abstraction (probably a separate database) that can understand a batch of jobs and track how many jobs are remaining in the "batch". Using separate queues for this is not a terrible approach, to be honest, but an abstraction will play more nicely with workers and things over time, I think.

It would be interesting to read up on when to use or not use a distinct queue.

Reply:
> This would facilitate the fat-finger check

The way I built the UI it should not be possible to double-click the send button and the user has to reload the page before they can get back to sending another email. So we should be ok on this front.